core/runtime/register_trt_op.cpp

```diff
@@ -10,80 +10,88 @@ namespace trtorch {
     return true;
   }
 
   // GPU case
   if (conf_device.device_type == nvinfer1::DeviceType::kGPU) {
     if (curr_device.device_name != conf_device.device_name) {
       LOG_WARNING(
-          "TRTEngine compiled for " << conf_device.device_name << " but current CUDA device is "
-                                    << curr_device.device_name << ". Switching the device context");
+          "Program compiled for " << conf_device.device_name << " but current CUDA device is " << curr_device
+                                  << ". Switching the device context");
       return true;
     }
   }
 
-  if (curr_device.id != conf_device.id) {
-    LOG_WARNING(
-        "Configured Device ID: " << conf_device.id << " is different that current device ID: " << curr_device.id
-                                 << ". Switching context");
-    return true;
-  }
+  // REVIEW: This shouldnt be a reason to switch GPU
+  // if (curr_device.id != conf_device.id) {
+  //   LOG_WARNING(
+  //       "Configured Device ID: " << conf_device.id << " is different that current device ID: " << curr_device.id
+  //                                << ". Switching context");
+  //   return true;
+  // }
 
   return false;
 }
 
-int select_cuda_device(const CudaDevice& conf_device) {
-  int device_id = 0;
-  auto dla_supported = get_dla_supported_SM();
+CudaDevice select_cuda_device(const CudaDevice& conf_device) {
+  int64_t device_id = -1;
+  auto dla_supported = get_dla_supported_SMs();
+
+  auto device_list = get_available_device_list().get_devices();
 
-  auto device_list = cuda_device_list.instance().get_devices();
+  CudaDevice new_target_device;
 
   for (auto device : device_list) {
-    auto compute_cap = std::to_string(device.second.major) + "." + std::to_string(device.second.minor);
+    auto compute_cap = device.second.getSMCapability();
     // In case of DLA select the DLA supported device ID
     if (conf_device.device_type == nvinfer1::DeviceType::kDLA) {
       if (dla_supported.find(compute_cap) != dla_supported.end() &&
           dla_supported[compute_cap] == device.second.device_name) {
         device_id = device.second.id;
+        new_target_device = CudaDevice(device_id, nvinfer1::DeviceType::kDLA);
         break;
       }
     } else if (conf_device.device_type == nvinfer1::DeviceType::kGPU) {
-      auto conf_sm = std::to_string(conf_device.major) + "." + std::to_string(conf_device.minor);
+      auto conf_sm = conf_device.getSMCapability();
       if (compute_cap == conf_sm && device.second.device_name == conf_device.device_name) {
         device_id = device.second.id;
+        new_target_device = CudaDevice(device_id, nvinfer1::DeviceType::kGPU);
         break;
       }
     } else {
-      LOG_ERROR("Unkown device type detected from the compiled engine");
+      TRTORCH_THROW_ERROR("Unknown target device type detected from the compiled program (runtime.select_cuda_device)");
```

Comment:
the LOG_ERROR would not stop the program from executing, not sure if we should continue if we dont know the device type